Setting up a development environment in a Dev Container

BrooklynDewolf opened the pull request:
## Changes introduced with this PR

* Included all the essential files to setup a Dev Container with Docker
* Included a guide `DEVELOPMENT.adoc` to setup a development environment in a Dev Container easily

Co-authored-by: Jean-Louis Dupond <jean-louis@dupond.be>

antonios-f committed:
Change save/restore nvram data logic

Save NVRAM data not only if BiosType is Q35_SECURE_BOOT
Also save if Q35_OVMF because it also includes EFI boot variables
With Q35_OVMF we unable to boot non-fallback bootloaders
in place different from <esp>/EFI/BOOT/BOOT<arch>.EFI

To check this it needed to:
1) Run VM with installed Linux (it doesn't matter what distro)
2) Add a bootloader via efibbotmgr
   efibootmgr -c -d <boot device> -L "Some Linux" -l \\EFI\\<somevendor>\\grubx64.efi
3) Look at created efi boot variable
   efibootmgr -v
   It's present
4) Shutdown VM
5) Run it again
6) Check efi boot variables again
   efibootmgr -v
7) We got non bootable VM if fallback bootloader is broken or boot default esp/EFI/BOOT/BOOT<arch>.EFI
   otherwise

Signed-off-by: Anton Fadeev <anton.fadeev@red-soft.ru>

shubhaOracle committed:
During VM import from SD, exclude shared disk when iterating over disk snapshots. Signed-off-by: Shubha Kulkarni shubha.kulkarni@oracle.com

Signed-off-by: ShubhaOracle <Shubha.kulkarni@oracle.com>

B4N4N41C committed:
Added a warning window about clearing NVRAM when changing Bios type and added a condition in the Remove_nvram_data function

When changing the chipset from the Q35 chipset with UEFI SecureBoot to any other and with a Q35 chipset with UEFI, a Q35 chipset with BIOS, or an I440FX chipset with BIOS, the nvram is cleared. A window based on the ConfirmationModel has been added to notify the user about clearing NVRAM when changing the BIOS type

Signed-off-by: Mochalin Nikolay <mochalin.nikolay2017@yandex.ru>

Commit:
Rewrite code to correspond with general oVirt style

Signed-off-by: Mochalin Nikolay <nikolay.mochalin@red-soft.ru>

BrooklynDewolf committed:
Included all the essential files to setup a Dev Container with Docker and included a guide DEVELOPMENT.adoc to setup dev environment easily

All the instructions on how to setup the Dev Container environment can be found in DEVELOPMENT.adoc

Co-authored-by: Jean-Louis Dupond <jean-louis@dupond.be>
Signed-off-by: Brooklyn Dewolf <contact@brooklyn.gent>

dupondje committed:
Fix CA generation as non-root user due to .rnd error

engine-setup fails creating the CA certificate when running under non-root user with the following error:

---------

plugin.execute:923 execute-output: ('/home/build/**FILTERED**//share/**FILTERED**-engine/bin/pki-create-ca.sh', '--subject=/C=US/O=Test/CN=c0714690d92b.71630', '--keystore-password=**FILTERED**', '--ca-file=ca') stderr:
Can't load .rnd into RNG
803BCAC12B7F0000:error:12000079:random number generator:RAND_load_file:Cannot open file:crypto/rand/randfile.c:106:Filename=.rnd
Cannot write random bytes:
803BCAC12B7F0000:error:12000079:random number generator:RAND_write_file:Cannot open file:crypto/rand/randfile.c:240:Filename=.rnd
Cannot generate CA request

----------

openssl commands try to find a .rnd file in the current directly. If not found, it will be created.
But as we do not change early enough into the correct path, the .rnd file can't be created there, resulting in an error.
So we just switch move the openssl req command in the subshell in the correct PKIDIR path.

Signed-off-by: Brooklyn Dewolf <contact@brooklyn.gent>
Signed-off-by: Jean-Louis Dupond <jean-louis@dupond.be>

0ffer committed:
engine: add scope 'openid' for compatibility with new versions of the keycloak

This is linked with this change for V19 and higher: https://www.keycloak.org/docs/latest/upgrading/index.html#userinfo-endpoint-changes

Now for request user-info endpoint we need request token with 'openid' scope (this is required scope by standard)

Signed-off-by: Melnichuk Stas <melnichuk.stas@gmail.com>

dupondje commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,9 @@
+  "build": {
+    "dockerfile": "../Dockerfile"
+  },
+  "runArgs": ["--network=ovirt", "--ulimit=nofile=2048:2048"],
```
Document why the ulimit is needed

dupondje commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,9 @@
+  },
+  "runArgs": ["--network=ovirt", "--ulimit=nofile=2048:2048"],
+  "mounts": ["source=ovirt-pki,target=/home/build/ovirt/etc/pki/ovirt-engine,type=volume"],
+  "appPort": [ "8080:8080", "8443:8443", "8787:8787", "54323:54323", "9696:9696", "6642:6642", "35357:35357", "2222:2222", "6100:6100", "7410:7410"]
```
Might be nice to put in some comment for which service each port is needed?

dupondje commented on DEVELOPMENT.adoc:
```diff
@@ -0,0 +1,75 @@
+# oVirt Engine - Dev Container setup & building
+
+Before beginning, ensure that [Docker](https://www.docker.com/) is installed on the machine that the Dev Container will be running on. This could be your local device or on a [remote machine](https://code.visualstudio.com/docs/devcontainers/containers#_open-a-folder-on-a-remote-ssh-host-in-a-container).
```
s/Before beginning/Before starting/

dupondje commented on DEVELOPMENT.adoc:
```diff
@@ -0,0 +1,75 @@
+
+Before beginning, ensure that [Docker](https://www.docker.com/) is installed on the machine that the Dev Container will be running on. This could be your local device or on a [remote machine](https://code.visualstudio.com/docs/devcontainers/containers#_open-a-folder-on-a-remote-ssh-host-in-a-container).
+
+### Setup the PostgresQL database
```
PostgreSQL

dupondje commented on DEVELOPMENT.adoc:
````diff
@@ -0,0 +1,75 @@
+make install-dev PREFIX=/home/build/ovirt/
+```
+
+Running this command will initiate the application's build process. You can skip tests by appending the `SKIP_TESTS=1` flag to the previous command. Additionally, you can specify the installation directory using the `PREFIX` flag.
````
env vars are always set at the beginning of the command normally :)
SO: `SKIP_TESTS=1 make install-dev PREFIX=/home/build/ovirt/`

dupondje commented on DEVELOPMENT.adoc:
````diff
@@ -0,0 +1,75 @@
+ovirt-imageio --conf-dir /home/build/ovirt/etc/ovirt-imageio
+```
+
+> If encountering a certificate error during connection testing, you'll need to accept the certificate in your browser. You can either locate the URL by accessing your browser's network console, clicking the 'Test Connection' button, and visiting the URL showing the error, OR simply navigate to `https://HOSTNAME:54323/info/` and accept the certificate.
````
Or you install the oVirt generated CA as a trusted CA in your browser

dupondje commented on DEVELOPMENT.adoc:
```diff
@@ -0,0 +1,75 @@
+
+### Connecting via console to your VM instance
+
+To establish a connection to the VM console via noVCN, you'll need to initiate the websocket service:
```
noVNC

dupondje commented on DEVELOPMENT.adoc:
````diff
@@ -0,0 +1,75 @@
+/home/build/ovirt/share/ovirt-engine/services/ovirt-websocket-proxy/ovirt-websocket-proxy.py start
+```
+
+Afterward, head to your newly created VM and select the 'v' arrow adjacent to the 'Console' button. Proceed to 'Console Options' and set 'Console Invocation' to 'noVCN'. Save these settings and click on the 'Console' button.
````
Afterwards
noVNC

dupondje commented on frontend/webadmin/modules/pom.xml:
```diff
@@ -43,7 +43,7 @@
     <!-- Control target browsers for GWT compilation, maps to 'user.agent' deferred binding property in *.gwt.xml -->
     <!-- By default, compile for Firefox browser only, use 'all-user-agents' profile to compile for all browsers -->
-    <gwt.userAgent>gecko1_8</gwt.userAgent>
+    <gwt.userAgent>safari</gwt.userAgent>
```
Unrelated, remove change

dupondje commented on packaging/bin/pki-create-ca.sh:
Would put this change in another PR

BrooklynDewolf commented:
All feedback should have been processed, looks good?

Commit:
Simplification of conditions

Signed-off-by: Mochalin Nikolay <nikolay.mochalin@red-soft.ru>

antonios-f committed:
Update SaveVmExternalDataCommand.java

Fix typo in method name.
isUEFI -> isUefi

Signed-off-by: Anton Fadeev <anton.fadeev@red-soft.ru>

antonios-f committed:
Fix checkstyle violations in VmListModel.java

Signed-off-by: Anton Fadeev <anton.fadeev@red-soft.ru>

antonios-f committed:
Fix comparision in VmListModel.java

Signed-off-by: Anton Fadeev <anton.fadeev@red-soft.ru>

0ffer committed:
Handle situation when empty groups not come from the keycloak (and hence from mod_auth_openidc)

Before the keycloak v22 it sends empty array as group claim when user not a member of any group.
After v22 it not put this claim at all.
Look discussion: https://github.com/keycloak/keycloak/issues/22340

Signed-off-by: Stanislav Melnichuk <melnichuk.stas@gmail.com>

0ffer committed:
db: cleanup: Remove crumbs after materialized views deletion

There were some remained parts after materialized views functionality deletion.
Comments and if-blocks that don't mean anything now.
Initial deletion commit hash: fa5ead39d36cb999460f853d4573a36dba8ec8b0

Signed-off-by: Stanislav Melnichuk <melnichuk.stas@gmail.com>

dupondje committed:
Create VmBackup earlier in HybridBackupCommand

When creation of a snapshot for the HybridBackup fails, we end up with a
null pointer exception.

2024-04-17 10:29:48,296+02 ERROR [org.ovirt.engine.core.dal.dbbroker.auditloghandling.AuditLogDirector] (default task-2) [76e81a20-4d1d-441b-8e6c-90f37889cd66] EVENT_ID: USER_FAILED_CREATE_SNAPSHOT(117), Failed to create Snapshot Auto-generated for Backup VM for VM xxxxx (User: admin@internal-authz).
2024-04-17 10:29:48,298+02 INFO  [org.ovirt.engine.core.bll.storage.backup.HybridBackupCommand] (default task-2) [76e81a20-4d1d-441b-8e6c-90f37889cd66] Change VM '73df70f5-09b0-4f1a-abd8-2d0b4cd5e8b2' backup 'null' phase from 'null' to 'FAILED'
2024-04-17 10:29:48,299+02 ERROR [org.ovirt.engine.core.bll.storage.backup.HybridBackupCommand] (default task-2) [76e81a20-4d1d-441b-8e6c-90f37889cd66] Command 'org.ovirt.engine.core.bll.storage.backup.HybridBackupCommand' failed: null
2024-04-17 10:29:48,299+02 ERROR [org.ovirt.engine.core.bll.storage.backup.HybridBackupCommand] (default task-2) [76e81a20-4d1d-441b-8e6c-90f37889cd66] Exception: java.lang.NullPointerException
	at org.ovirt.engine.core.dal//org.ovirt.engine.core.dao.VmBackupDaoImpl.update(VmBackupDaoImpl.java:90)
	at org.ovirt.engine.core.dal//org.ovirt.engine.core.dao.VmBackupDaoImpl.update(VmBackupDaoImpl.java:22)
	at deployment.engine.ear.bll.jar//org.ovirt.engine.core.bll.storage.backup.StartVmBackupCommand.updateVmBackupPhase(StartVmBackupCommand.java:632)
	at deployment.engine.ear.bll.jar//org.ovirt.engine.core.bll.storage.backup.HybridBackupCommand.executeCommand(HybridBackupCommand.java:118)
	at deployment.engine.ear.bll.jar//org.ovirt.engine.core.bll.CommandBase.executeWithoutTransaction(CommandBase.java:1174)
	at deployment.engine.ear.bll.jar//org.ovirt.engine.core.bll.CommandBase.executeActionInTransactionScope(CommandBase.java:1332)

This happens because VmBackupDaoImpl tries to save the BackupType into
the database (entity.getBackupType().getName()), but that value is still
null.
The BackupType is only set when calling createVmBackup.

But as we already have all the data except the snapshotId at the start
of the HybridBackup, we already call createVmBackup before creating the
snapshot and when the snapshot was created, we add it to the VmBackup
data.

This avoids the null getBackupType()

Signed-off-by: Jean-Louis Dupond <jean-louis@dupond.be>

Commit:
Add support for DC/cluster level 4.8

This patch adds support for DC/cluster compatibility level 4.8. Hosts need to be running on libvirt >= 9.5 and qemu-kvm 8.1.

Signed-off-by: Brooklyn Dewolf <contact@brooklyn.gent>

Commit:
Added discard-no-unref support to CL 4.8

The discard-no-unref was introduced to qemu to counteract fragmentation. Without this, qcow2 images would grow over 100% of their size. See https://patchew.org/QEMU/20230602124747.1544077-1-jean-louis@dupond.be/

I have added the option to ovirt-engine so that this flag is enabled by default on hosts with cluster level compatibility level 4.8. The EnableQemuDiscardNoUnref option can be enabled/disabled in the config. The Virtual Disk option 'Enable Discard' also needs to be active before this flag is enabled.

Signed-off-by: Brooklyn Dewolf <contact@brooklyn.gent>

dupondje committed:
packaging: fix error when enabling KeyCloak

This fixes engine-setup error:
[ ERROR ] Failed to execute stage 'Misc configuration': 'OVESETUP_OVN/ovirtProviderOvnSecret'

Since commit 978c90e we save the ovirtProviderOvnSecret value in the
setup file.
But if this value is not there, but ovirtProviderOvn is True, the
engine-setup fails.

So if we can't find the value in the config, a new password is
generated.

Signed-off-by: Jean-Louis Dupond <jean-louis@dupond.be>

Commit:
Bump machine type for cluster level 4.8

Hosts supporting level 4.8 need to be running on RHEL 9.4 or CentOS Stream/RHEL 9.

Signed-off-by: Brooklyn Dewolf <contact@brooklyn.gent>

0ffer committed:
engine: Check concrete disk permissions firstly on TransferImageStatusCommand

There is some situations when storageDomainId is not come with parameters (image upload cancel and pause operations). For this operations checked CREATE_DISK permission on SYSTEM_OBJECT (i.e. system-wide).
Problem starts when we give permissions for user only on concrete storage domain object (not system-wide). Then permission check failed for operations without storage domain id info in parameters. Here I just add check permission for disk before other objects.

Signed-off-by: Stanislav Melnichuk <melnichuk.stas@gmail.com>

sandrobonazzola committed:
automation: drop el8 after centos stream 8 eol

Dropping automation on el8 stream as centos stream 8 reached EOL.

Signed-off-by: Sandro Bonazzola <sandro.bonazzola@gmail.com>

Commit:
engine: Fix duplicate VM backup entry error when creating incremental backup

Fixes an issue that originates from https://github.com/oVirt/ovirt-engine/commit/ed023e5e0a000029b4d7e50b03fdbb3fd9a773bd

 The following error occured when trying to create an incremental backup:

 ------
 ERROR [org.ovirt.engine.core.bll.storage.backup.HybridBackupCommand] (default task-1) [full_cold_vm_backup] Command 'org.ovirt.engine.core.bll.storage.backup.HybridBackupCommand' failed: CallableStatementCallback; SQL [{call insertvmbackup(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}]; ERROR: duplicate key value violates unique constraint "vm_backups_pkey"
   Detail: Key (backup_id)=(1733241d-4a0b-4205-9e52-5383eb7a82f2) already exists.
 -----

 This was caused by using vmBackupDao.save() instead of vmBackupDao.update(), because the backup already existed when the setVmBackupSnapshot method is executed.

Signed-off-by: Brooklyn Dewolf <contact@brooklyn.gent>
